From 8eba8f03d027153830485ffd3b8bf69e0b3f949e Mon Sep 17 00:00:00 2001
From: Andrew Bays <andrew.bays@gmail.com>
Date: Fri, 14 Jun 2019 13:01:59 -0400
Subject: [PATCH 38/60] Bump CI for connectivity plugin

---
 src/connectivity.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/connectivity.c b/src/connectivity.c
index 3b160e76..204bbf65 100644
--- a/src/connectivity.c
+++ b/src/connectivity.c
@@ -21,7 +21,7 @@
  *
  * Authors:
  *   Red Hat NFVPE
- *     Andrew Bays <abays at redhat.com>
+ *     Andrew Bays <abays at redhat.com> 
  *     Aneesh Puttur <aputtur at redhat.com>
  **/
 
-- 
2.21.0

