From 5edb45ef1919daad5a366220b383721625672ad2 Mon Sep 17 00:00:00 2001
From: Andrew Bays <abays@redhat.com>
Date: Tue, 27 Feb 2018 13:18:05 -0500
Subject: [PATCH 48/60] Only declare rsyslog keys if yajl2 present

---
 src/sysevent.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/sysevent.c b/src/sysevent.c
index 94e99e9a..6ef79f95 100644
--- a/src/sysevent.c
+++ b/src/sysevent.c
@@ -119,10 +119,12 @@ static int buffer_length = 10;
 
 static int monitor_all_messages = 1;
 
+#if HAVE_YAJL_V2
 static const char *rsyslog_keys[3] = {"@timestamp", "@source_host", "@message"};
 static const char *rsyslog_field_keys[5] = {
     "facility", "severity", "severity-num", "program", "processid"};
-
+#endif
+    
 /*
  * Private functions
  */
-- 
2.21.0

