From c7f0935a1f66ecb57fb261ea4323d376aa12392c Mon Sep 17 00:00:00 2001
From: Andrew Bays <abays@redhat.com>
Date: Tue, 27 Feb 2018 14:05:14 -0500
Subject: [PATCH 49/60] clang formatting

---
 src/sysevent.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/sysevent.c b/src/sysevent.c
index 6ef79f95..ab4571b0 100644
--- a/src/sysevent.c
+++ b/src/sysevent.c
@@ -124,7 +124,7 @@ static const char *rsyslog_keys[3] = {"@timestamp", "@source_host", "@message"};
 static const char *rsyslog_field_keys[5] = {
     "facility", "severity", "severity-num", "program", "processid"};
 #endif
-    
+
 /*
  * Private functions
  */
-- 
2.21.0

