From 63f07548ca7790cd604f935783ec75a9c927edec Mon Sep 17 00:00:00 2001
From: Andrew Bays <abays@redhat.com>
Date: Wed, 28 Feb 2018 07:06:00 -0500
Subject: [PATCH 50/60] Remove bogus include

---
 src/sysevent.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/sysevent.c b/src/sysevent.c
index ab4571b0..49efe75c 100644
--- a/src/sysevent.c
+++ b/src/sysevent.c
@@ -31,7 +31,6 @@
 #include "utils_complain.h"
 #include "utils_ignorelist.h"
 
-#include <asm/types.h>
 #include <errno.h>
 #include <netdb.h>
 #include <netinet/in.h>
-- 
2.21.0

