From 1aed437355521e5740a90699c75a984768f1881f Mon Sep 17 00:00:00 2001
From: Matthias Runge <mrunge@redhat.com>
Date: Wed, 19 Jun 2019 12:16:10 +0200
Subject: [PATCH 61/61] Fix rebase issue

---
 configure.ac | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index 58336a11..085dd2f4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6488,17 +6488,12 @@ if test "x$ac_system" = "xLinux"; then
   if test "x$with_libyajl" = "xyes" && test "x$with_libyajl2" = "xyes"; then
     plugin_ovs_events="yes"
     plugin_ovs_stats="yes"
-<<<<<<< HEAD
     plugin_procevent="yes"
 
     if test "x$with_libmnl" = "xyes"; then
       plugin_connectivity="yes"
     fi
     plugin_sysevent="yes"
-||||||| merged common ancestors
-    plugin_sysevent="yes"
-=======
->>>>>>> VES-enabled sysvent plugin
   fi
 fi
 
@@ -6889,7 +6884,6 @@ AC_PLUGIN([pcie_errors],         [$plugin_pcie_errors],       [PCIe errors plugi
 AC_PLUGIN([perl],                [$plugin_perl],              [Embed a Perl interpreter])
 AC_PLUGIN([pf],                  [$have_net_pfvar_h],         [BSD packet filter (PF) statistics])
 # FIXME: Check for libevent, too.
-<<<<<<< HEAD
 AC_PLUGIN([pinba],               [$plugin_pinba],             [Pinba statistics])
 AC_PLUGIN([ping],                [$with_liboping],            [Network latency statistics])
 AC_PLUGIN([postgresql],          [$with_libpq],               [PostgreSQL database statistics])
-- 
2.21.0

